homeassistant/components/zwave_js/__init__.py

```diff
@@ -55,49 +62,70 @@ async def async_on_initialized():
+            if not node.ready:
+                continue
+            asyncio.create_task(async_on_node_ready(node))
+        # TODO: register callback for "node_ready" event
 
     async def async_on_disconnect():
         """Handle websocket is disconnected."""
         LOGGER.info("Disconnected from Zwave JS Server")
         # TODO: signal entities to update availability state
 
-    for component in PLATFORMS:
-        hass.async_create_task(
-            hass.config_entries.async_forward_entry_setup(entry, component)
-        )
+    async def async_on_node_ready(node: ZwaveNode) -> None:
+        """Handle node ready event."""
+        LOGGER.debug("Processing node %s", node)
+        # register (or update) node in device registry
+        dev_reg = await device_registry.async_get_registry(hass)
+        register_node_in_dev_reg(entry, dev_reg, client, node)
+        # run discovery on all node values and create/update entities
+        async for disc_info in async_discover_values(node):
+            LOGGER.debug("Discovered entity: %s", disc_info)
+            if disc_info.value_id not in discovered_values:
+                # dispatch discovery_info to platform
+                discovered_values.add(disc_info.value_id)
+                async_dispatcher_send(hass, f"{DOMAIN}_add_{disc_info.platform}", disc_info)
+            else:
+                # already discovered, dispatch update request
+                async_dispatcher_send(
+                    hass, f"{DOMAIN}_update_{disc_info.value_id}", disc_info.primary_value
+                )
 
     # register main event callbacks.
-    unsubs = [
-        client.register_on_initialized(async_on_initialized),
-        client.register_on_disconnect(async_on_disconnect),
-    ]
-
-    asyncio.create_task(client.connect())
-
-    try:
-        async with timeout(10):
-            await initialized.wait()
-    except asyncio.TimeoutError as err:
-        for unsub in unsubs:
-            unsub()
-        await client.disconnect()
-        raise ConfigEntryNotReady from err
+    client.register_on_initialized(async_on_initialized)
+    client.register_on_disconnect(async_on_disconnect)
+
+    async def async_connect():
+        """Start platforms and connect."""
+        # start platforms
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_setup(entry, platform)
+                for platform in PLATFORMS
+            ]
+        )
+        # connect and throw error if connection failed
+        await client.connect()
+        try:
+            async with timeout(10):
+                await initialized.wait()
+        except asyncio.TimeoutError as err:
+            for unsub in hass.data[DOMAIN][entry.entry_id][DATA_UNSUBSCRIBE]:
+                unsub()
+            await client.disconnect()
+            raise ConfigEntryNotReady from err
+
+    asyncio.create_task(async_connect())
```

Comment:
We don't want all to be here. Now `ConfigEntryNotReady` doesn't propagate correctly.

Reply:
The connect code needs to be inside `async_setup`.

Reply:
We also shouldn't forward the entries before we connect, as we might not continue setup if connection fails !

Reply:
well, that brings us into the race condition again... ideas ?
can we queue the dispatcher commands for adding the discover infos ?